src/services/extra_data.py

```diff
@@ -0,0 +1,129 @@
+        return OracleReportLimits(**dict(zip(output_names, result)))
+
+    def to_bytes(self, extra_data: list[ExtraDataItem]) -> bytes:
+        extra_data_bytes = b''
+        for item in extra_data:
+            extra_data_bytes += item.item_index
+            extra_data_bytes += item.item_type.value.to_bytes(Lengths.ITEM_TYPE)
+            extra_data_bytes += item.item_payload.module_id
+            extra_data_bytes += item.item_payload.node_ops_count
+            extra_data_bytes += item.item_payload.node_operator_ids
+            extra_data_bytes += item.item_payload.vals_counts
+        return self.w3.keccak(extra_data_bytes)
+
+    def build_extra_data(self, stucked_payloads: list[ItemPayload], exited_payloads: list[ItemPayload]):
+        index = 0
+        extra_data = []
+        for item in stucked_payloads:
+            extra_data.append(ExtraDataItem(
+                item_index=index.to_bytes(Lengths.ITEM_INDEX),
+                item_type=ItemType.EXTRA_DATA_TYPE_STUCK_VALIDATORS,
+                item_payload=item
+            ))
+            index += 1
+        for item in exited_payloads:
+            extra_data.append(ExtraDataItem(
+                item_index=index.to_bytes(Lengths.ITEM_INDEX),
+                item_type=ItemType.EXTRA_DATA_TYPE_EXITED_VALIDATORS,
+                item_payload=item
+            ))
+            index += 1
+        return extra_data
+
+    def build_validators_payloads(self, validators: list[LidoKey], modules: list[ContractModule], max_list_items_count) -> list[ItemPayload]:
+        payloads = []
+        module_to_operators = {}
+
+        for module in modules:
+            module_to_operators[module.stakingModuleAddress] = {"module": module, "operators": defaultdict(list)}
+        for validator in validators:
+            module_to_operators[validator.moduleAddress]["operators"][validator.operatorIndex].append(validator)
+
+        for mapping in sorted(module_to_operators.values(), key=lambda x: x["module"].id):
+            module = mapping["module"]
+            operators = mapping["operators"]
+            operators = sorted(operators.items(), key=itemgetter(0))
+            for chunk in chunks(operators, max_list_items_count):
+                operator_ids = []
+                vals_count = []
+                for operator_id, validators in chunk:
+                    operator_ids.append(operator_id.to_bytes(Lengths.NODE_OPERATOR_IDS))
+                    vals_count.append(len(validators).to_bytes(Lengths.STUCK_VALS_COUNTS))
+                payloads.append(ItemPayload(
+                    module_id=module.id.to_bytes(Lengths.MODULE_ID),
+                    node_ops_count=len(chunk).to_bytes(Lengths.NODE_OPS_COUNT),
+                    node_operator_ids=b"".join(operator_ids),
+                    vals_counts=b"".join(vals_count),
+                ))
+        return payloads
```

Comment:
there is an extra check for amount of items in one group in the contract 

https://github.com/lidofinance/lido-dao/blob/feature/shapella-upgrade/contracts/0.8.9/sanity_checks/OracleReportSanityChecker.sol#L403-L423

https://discord.com/channels/761182643269795850/1029294901487214593/1075778491347382293